In [47]:
import json
import numpy as np
import pandas as pd
import requests
import numpy
from pandas.io.json import json_normalize
import matplotlib.pyplot as plt

# Research Question: Find the top 10 players with the highest ratio between completed passes and attempted passes.

### Import the players dataset

In [48]:

players=pd.read_json("C:\\Users\\Debodeep\\Documents\\Sapienza Learning Materials\\ADM\\ADMHw2\\players.json")
players
#Displaying maximum number
pd.set_option('display.max_columns', 14)
players

passportArea  weight  \
0     {'name': 'Turkey', 'id': '792', 'alpha3code': ...      78   
1     {'name': 'Senegal', 'id': '686', 'alpha3code':...      73   
2     {'name': 'France', 'id': '250', 'alpha3code': ...      72   
3     {'name': 'Senegal', 'id': '686', 'alpha3code':...      82   
4     {'name': 'France', 'id': '250', 'alpha3code': ...      84   
...                                                 ...     ...   
3598  {'name': 'Tunisia', 'id': 788, 'alpha3code': '...      72   
3599  {'name': 'Peru', 'id': 604, 'alpha3code': 'PER...      76   
3600  {'name': 'Peru', 'id': 604, 'alpha3code': 'PER...      78   
3601  {'name': 'Morocco', 'id': 504, 'alpha3code': '...      70   
3602  {'name': 'Panama', 'id': 591, 'alpha3code': 'P...       0   

              firstName middleName             lastName currentTeamId  \
0                 Harun                           Tekin          4502   
1                Malang                            Sarr          3775   
2                  Over                        Mandanda          3772   
3     Alfred John Momar                         N'Diaye           683   
4              Ibrahima                     Konat\u00e9          2975   
...                 ...        ...                  ...           ...   
3598                Ali                    Ma\u00e2loul         16041   
3599     Carlos Alberto             C\u00e1ceda Oyaguez         15591   
3600  Miguel Gianpierre                   Araujo Blanco         12072   
3601         Ahmed Reda                       Tagnaouti         16183   
3602            Ricardo                   Guardia Avila         62943   

       birthDate  height                                               role  \
0     1989-06-17     187  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
1     1999-01-23     182  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
2     1998-10-26     176  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
3     1990-03-06     187  {'code2': 'MD', 'code3': 'MID', 'name': 'Midfi...   
4     1999-05-25     192  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
...          ...     ...                                                ...   
3598  1990-01-01     175  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
3599  1991-09-27     183  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
3600  1994-10-24     179  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
3601  1996-04-05     182  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
3602  1997-02-04       0  {'code2': 'MD', 'code3': 'MID', 'name': 'Midfi...   

                                              birthArea    wyId   foot  \
0     {'name': 'Turkey', 'id': '792', 'alpha3code': ...   32777  right   
1     {'name': 'France', 'id': '250', 'alpha3code': ...  393228   left   
2     {'name': 'France', 'id': '250', 'alpha3code': ...  393230          
3     {'name': 'France', 'id': '250', 'alpha3code': ...   32793  right   
4     {'name': 'France', 'id': '250', 'alpha3code': ...  393247  right   
...                                                 ...     ...    ...   
3598  {'name': 'Tunisia', 'id': 788, 'alpha3code': '...  120839   left   
3599  {'name': 'Peru', 'id': 604, 'alpha3code': 'PER...  114736  right   
3600  {'name': 'Peru', 'id': 604, 'alpha3code': 'PER...  114908  right   
3601  {'name': 'Morocco', 'id': 504, 'alpha3code': '...  285583  right   
3602  {'name': 'Panama', 'id': 591, 'alpha3code': 'P...  361536   left   

            shortName currentNationalTeamId  
0            H. Tekin                  4687  
1             M. Sarr                  4423  
2         O. Mandanda                  null  
3          A. N'Diaye                 19314  
4      I. Konat\u00e9                  null  
...               ...                   ...  
3598  A. Ma\u00e2loul                  null  
3599   C. C\u00e1ceda                  null  
3600        M. Araujo                  null  
3601     A. Tagnaouti                  null  
3602         R. A

### Import the event data for England

In [63]:
events=pd.read_json("C:\\Users\\Debodeep\\Documents\\Sapienza Learning Materials\\ADM\\ADMHw2\\events\\events_England.json")
pd.set_option('display.max_columns', 15)
events

eventId           subEventName  \
0             8            Simple pass   
1             8              High pass   
2             8              Head pass   
3             8              Head pass   
4             8            Simple pass   
...         ...                    ...   
643145        5  Ball out of the field   
643146        3                 Corner   
643147        1               Air duel   
643148        1               Air duel   
643149       10                   Shot   

                                                     tags  playerId  \
0                                          [{'id': 1801}]     25413   
1                                          [{'id': 1801}]    370224   
2                                          [{'id': 1801}]      3319   
3                                          [{'id': 1801}]    120339   
4                                          [{'id': 1801}]    167145   
...                                                   ...       ...   
643145                                                 []         0   
643146           [{'id': 302}, {'id': 801}, {'id': 1801}]     70965   
643147                        [{'id': 701}, {'id': 1802}]      7919   
643148                        [{'id': 703}, {'id': 1801}]      8005   
643149  [{'id': 403}, {'id': 201}, {'id': 1216}, {'id'...      8005   

                                         positions  matchId     eventName  \
0         [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  2499719          Pass   
1         [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  2499719          Pass   
2         [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  2499719          Pass   
3         [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  2499719          Pass   
4         [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  2499719          Pass   
...                                            ...      ...           ...   
643145   [{'y': 32, 'x': 0}, {'y': 100, 'x': 100}]  2500098  Interruption   
643146  [{'y': 100, 'x': 100}, {'y': 47, 'x': 88}]  2500098     Free Kick   
643147    [{'y': 53, 'x': 12}, {'y': 50, 'x': 14}]  2500098          Duel   
643148    [{'y': 47, 'x': 88}, {'y': 50, 'x': 86}]  2500098          Duel   
643149      [{'y': 50, 'x': 86}, {'y': 0, 'x': 0}]  2500098          Shot   

        teamId matchPeriod     eventSec subEventId         id  
0         1609          1H     2.758649         85  177959171  
1         1609          1H     4.946850         83  177959172  
2         1609          1H     6.542188         82  177959173  
3         1609          1H     8.143395         82  177959174  
4         1609          1H    10.302366         85  177959175  
...        ...         ...          ...        ...        ...  
643145    1623          2H  2796.732525         50  251596409  
643146    1633          2H  2829.821084         30  251596232  
643147    1623          2H  2831.211419         10  251596410  
643148    1633          2H  2832.434399         10  251596234  
643149    1633          2H  2834.127168        100  251596236  

[643150 rows x 12 columns]

#### As we have to calculate the efficiency of the players based on the passes, we filter out the information related to passes. Thus we target the column 'eventName' and choose only those rows which contains information related to passes. 

In [50]:
events['eventName']=events['eventName'].astype(str)
events=events[(events.eventName.str.contains('Pass'))]
pd.set_option('display.max_columns', 15)
events

eventId subEventName                                       tags  \
0             8  Simple pass                             [{'id': 1801}]   
1             8    High pass                             [{'id': 1801}]   
2             8    Head pass                             [{'id': 1801}]   
3             8    Head pass                             [{'id': 1801}]   
4             8  Simple pass                             [{'id': 1801}]   
...         ...          ...                                        ...   
643134        8  Simple pass                             [{'id': 1801}]   
643135        8  Simple pass                             [{'id': 1801}]   
643141        8  Simple pass                             [{'id': 1801}]   
643142        8    High pass                             [{'id': 1801}]   
643143        8        Cross  [{'id': 402}, {'id': 2101}, {'id': 1802}]   

        playerId                                 positions  matchId eventName  \
0          25413  [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  2499719      Pass   
1         370224  [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  2499719      Pass   
2           3319  [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  2499719      Pass   
3         120339  [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  2499719      Pass   
4         167145  [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  2499719      Pass   
...          ...                                       ...      ...       ...   
643134      8313  [{'y': 63, 'x': 65}, {'y': 96, 'x': 61}]  2500098      Pass   
643135    379209  [{'y': 96, 'x': 61}, {'y': 85, 'x': 40}]  2500098      Pass   
643141      8561  [{'y': 70, 'x': 59}, {'y': 89, 'x': 53}]  2500098      Pass   
643142     20620  [{'y': 89, 'x': 53}, {'y': 72, 'x': 99}]  2500098      Pass   
643143     14703    [{'y': 72, 'x': 99}, {'y': 0, 'x': 0}]  2500098      Pass   

        teamId matchPeriod     eventSec subEventId         id  
0         1609          1H     2.758649         85  177959171  
1         1609          1H     4.946850         83  177959172  
2         1609          1H     6.542188         82  177959173  
3         1609          1H     8.143395         82  177959174  
4         1609          1H    10.302366         85  177959175  
...        ...         ...          ...        ...        ...  
643134    1633          2H  2760.286946         85  251596219  
643135    1633          2H  2763.698599         85  251596221  
643141    1633          2H  2783.137836         85  251596225  
643142    1633          2H  2785.702032         83  251596226  
643143    1633          2H  2789.232666         80  251596229  

[328657 rows x 12 columns]

#### Splitting the column 'Tags' that is a dictionary

In [51]:
TAG1=[]
for i in range(len(events.tags)):
    try:
        TAG1.append(str(events.tags[i][0]['id']))
    except:
        TAG1.append('Notag')
TAG2=[]
for i in range(len(events.tags)):
    try:
        TAG2.append(str(events.tags[i][1]['id']))
    except:
        TAG2.append('Notag')
TAG3=[]
for i in range(len(events.tags)):
    try:
        TAG3.append(str(events.tags[i][2]['id']))
    except:
        TAG3.append('Notag')
TAG4=[]
for i in range(len(events.tags)):
    try:
        TAG4.append(str(events.tags[i][2]['id']))
    except:
        TAG4.append('Notag')
TAG1=pd.DataFrame({'Tag1':TAG1})
TAG2=pd.DataFrame({'Tag2':TAG2})
TAG3=pd.DataFrame({'Tag3':TAG3})
TAG4=pd.DataFrame({'Tag4':TAG4})
TAG=[TAG1,TAG2,TAG3,TAG4]
TAG=pd.concat(TAG ,axis=1, sort=False)
mix_events=[events,TAG]
events=pd.concat(mix_events,  axis=1,  sort=False).reindex(events.index)
events=events.drop('tags',axis=1)
events

eventId subEventName  playerId  \
0           8.0  Simple pass   25413.0   
1           8.0    High pass  370224.0   
2           8.0    Head pass    3319.0   
3           8.0    Head pass  120339.0   
4           8.0  Simple pass  167145.0   
...         ...          ...       ...   
643134      8.0  Simple pass    8313.0   
643135      8.0  Simple pass  379209.0   
643141      8.0  Simple pass    8561.0   
643142      8.0    High pass   20620.0   
643143      8.0        Cross   14703.0   

                                       positions    matchId eventName  teamId  \
0       [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  2499719.0      Pass  1609.0   
1       [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  2499719.0      Pass  1609.0   
2       [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  2499719.0      Pass  1609.0   
3       [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  2499719.0      Pass  1609.0   
4       [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  2499719.0      Pass  1609.0   
...                                          ...        ...       ...     ...   
643134  [{'y': 63, 'x': 65}, {'y': 96, 'x': 61}]  2500098.0      Pass  1633.0   
643135  [{'y': 96, 'x': 61}, {'y': 85, 'x': 40}]  2500098.0      Pass  1633.0   
643141  [{'y': 70, 'x': 59}, {'y': 89, 'x': 53}]  2500098.0      Pass  1633.0   
643142  [{'y': 89, 'x': 53}, {'y': 72, 'x': 99}]  2500098.0      Pass  1633.0   
643143    [{'y': 72, 'x': 99}, {'y': 0, 'x': 0}]  2500098.0      Pass  1633.0   

       matchPeriod     eventSec subEventId           id  Tag1   Tag2   Tag3  \
0               1H     2.758649         85  177959171.0  1801  Notag  Notag   
1               1H     4.946850         83  177959172.0  1801  Notag  Notag   
2               1H     6.542188         82  177959173.0  1801  Notag  Notag   
3               1H     8.143395         82  177959174.0  1801  Notag  Notag   
4               1H    10.302366         85  177959175.0  1801  Notag  Notag   
...            ...          ...        ...          ...   ...    ...    ...   
643134          2H  2760.286946         85  251596219.0   NaN    NaN    NaN   
643135          2H  2763.698599         85  251596221.0   NaN    NaN    NaN   
643141          2H  2783.137836         85  251596225.0   NaN    NaN    NaN   
643142          2H  2785.702032         83  251596226.0   NaN    NaN    NaN   
643143          2H  2789.232666         80  251596229.0   NaN    NaN    NaN   

         Tag4  
0       Notag  
1       Notag  
2       Notag  
3       Notag  
4       Notag  
...       ...  
643134    NaN  
643135    NaN  
643141    NaN  
643142    NaN  
643143    NaN  

[328657 rows x 15 columns]

#### Counting the number of passes with respect to the player id

In [52]:
passes=events
passes=passes.drop('positions',axis=1)
passes.playerId=list(passes.playerId)
PLAYERS=[]
for i in range(len(passes.playerId)):
    PLAYERS.append(passes.playerId.iloc[i])
passes_player=pd.DataFrame(PLAYERS , columns=['PlayerId'])
passes_player['Numberofpasses'] = passes_player.groupby('PlayerId')['PlayerId'].transform('count')
passes_player

PlayerId  Numberofpasses
0        25413.0             657
1       370224.0             574
2         3319.0            1790
3       120339.0             834
4       167145.0            1877
...          ...             ...
328652    8313.0            1203
328653  379209.0             549
328654    8561.0            1245
328655   20620.0             725
328656   14703.0             689

[328657 rows x 2 columns]

#### In order to set the threshold of the number of passes, we calculate the first quartile of the number of passes. Then we set the lower limit as (1st quartile*1.5). The reason behind this is, while dealing with a set of variables, the upper and lower boundary is being set as  (3rd quartile*1.5) and (1st quartile*1.5) respectively. Here we are concerned about the players who have played very limited number of matches or insufficient amount of hours. That is why we set the lower limit as the only threshold.

In [53]:
#calculating 1st quartile
quartile_1st = passes_player.Numberofpasses.quantile(0.25)
quartile_1st=int(quartile_1st)
#calculating lower bound
lower_bound=quartile_1st*1.5
ID_passes_threshold=[]
frequency_passes_threshold=[]
frequency_passes=[]
ID_passes=[]
for i in range(len(passes_player.Numberofpasses)):
    frequency_passes.append(float(passes_player.Numberofpasses.iloc[i]))
    ID_passes.append(float(passes_player.PlayerId.iloc[i]))
for i in range(len(frequency_passes)):
    if frequency_passes[i]>lower_bound:
        frequency_passes_threshold.append(int(frequency_passes[i]))
        ID_passes_threshold.append(int(ID_passes[i]))

In [54]:
passes_threshold=pd.DataFrame(ID_passes_threshold, columns=['PlayerId'])
passes_threshold.insert(1, "Numberofpasses",frequency_passes_threshold, True)
passes_threshold.drop_duplicates(subset='PlayerId', keep='first', inplace=True)
passes_threshold

PlayerId  Numberofpasses
0          3319            1790
1        167145            1877
3          8653            1640
5         49876            2974
6         14869            1251
...         ...             ...
20659     14808            1350
23436      8464            1584
23994      7922            1071
25254    302518            1184
28933       370            1635

[112 rows x 2 columns]

In [55]:
events_tag2=events.loc[events['Tag2'] == '1801']
events_tag1=events.loc[events['Tag1']=='1801']
events_tag3=events.loc[events['Tag3']=='1801']
events_tag4=events.loc[events['Tag4']=='1801']
events_tag=[events_tag1,events_tag2,events_tag3,events_tag4]
success=pd.concat(events_tag, axis=0)
success

eventId subEventName  playerId  \
0           8.0  Simple pass   25413.0   
1           8.0    High pass  370224.0   
2           8.0    Head pass    3319.0   
3           8.0    Head pass  120339.0   
4           8.0  Simple pass  167145.0   
...         ...          ...       ...   
328196      8.0        Cross    8976.0   
328283      8.0        Cross  346101.0   
328286      8.0        Cross   15808.0   
328426      8.0        Cross  346101.0   
328646      8.0        Cross   77557.0   

                                       positions    matchId eventName  \
0       [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  2499719.0      Pass   
1       [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  2499719.0      Pass   
2       [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  2499719.0      Pass   
3       [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  2499719.0      Pass   
4       [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  2499719.0      Pass   
...                                          ...        ...       ...   
328196  [{'y': 14, 'x': 95}, {'y': 54, 'x': 90}]  2499913.0      Pass   
328283  [{'y': 88, 'x': 89}, {'y': 30, 'x': 96}]  2499913.0      Pass   
328286  [{'y': 29, 'x': 96}, {'y': 57, 'x': 94}]  2499913.0      Pass   
328426  [{'y': 93, 'x': 78}, {'y': 52, 'x': 88}]  2499913.0      Pass   
328646  [{'y': 93, 'x': 91}, {'y': 61, 'x': 89}]  2499913.0      Pass   

         teamId matchPeriod     eventSec subEventId           id  Tag1   Tag2  \
0        1609.0          1H     2.758649         85  177959171.0  1801  Notag   
1        1609.0          1H     4.946850         83  177959172.0  1801  Notag   
2        1609.0          1H     6.542188         82  177959173.0  1801  Notag   
3        1609.0          1H     8.143395         82  177959174.0  1801  Notag   
4        1609.0          1H    10.302366         85  177959175.0  1801  Notag   
...         ...         ...          ...        ...          ...   ...    ...   
328196  10531.0          1H  2186.328315         80  217543225.0   401    801   
328283   1612.0          1H  2415.680872         80  217543075.0   402    801   
328286   1612.0          1H  2422.205856         80  217543079.0   401    801   
328426   1612.0          2H   129.777831         80  217543342.0   402    801   
328646  10531.0          2H   894.779661         80  217543713.0   302    402   

         Tag3   Tag4  
0       Notag  Notag  
1       Notag  Notag  
2       Notag  Notag  
3       Notag  Notag  
4       Notag  Notag  
...       ...    ...  
328196   1801   1801  
328283   1801   1801  
328286   1801   1801  
328426   1801   1801  
328646   1801   1801  

[141130 rows x 15 columns]

In [56]:
SUCCESS=[]
for i in range(len(success.playerId)):
    SUCCESS.append(success.playerId.iloc[i])
success_new=pd.DataFrame(SUCCESS , columns=['PlayerId'])
success_new['Numberofpasses'] = success_new.groupby('PlayerId')['PlayerId'].transform('count')
success_new.drop_duplicates(subset='PlayerId', keep='first', inplace=True)
success_new

PlayerId  Numberofpasses
0        25413.0             298
1       370224.0             185
2         3319.0            1012
3       120339.0             205
4       167145.0             937
...          ...             ...
124506  281135.0              31
124769  274488.0              17
126226   25604.0              17
127966  171283.0              48
137275    9179.0               1

[451 rows x 2 columns]

In [64]:
final_passes=pd.merge(success_new, passes_threshold, on='PlayerId')
final_passes=final_passes.rename(columns={'Numberofpasses_x':'AccuratePassesNumber','Numberofpasses_y':'TotalPassesNumber'})
final_passes

PlayerId  AccuratePassesNumber  TotalPassesNumber
0      3319.0                  1012               1790
1    167145.0                   937               1877
2      8653.0                   634               1640
3     49876.0                  1377               2974
4     14869.0                   688               1251
..        ...                   ...                ...
107   14808.0                   276               1350
108    8464.0                   970               1584
109    7922.0                   425               1071
110  302518.0                   354               1184
111     370.0                   520               1635

[112 rows x 3 columns]

In [65]:
def make_col(df, col):
    return pd.concat([df.drop([col], axis=1), df[col].apply(pd.Series)], axis=1)
teams = pd.read_json("C:\\Users\\Debodeep\\Documents\\Sapienza Learning Materials\\ADM\\ADMHw2\\teams.json").rename(columns={'name':'teamname'})
teams = make_col(teams, "area")
teams = teams.drop(columns='city')
teams = teams.rename(columns={'name':'nation'})
engteams = teams[(teams["alpha3code"]=="XEN") & (teams["type"]=="club") | (teams["teamname"]=="Swansea City")]
engplayer = players.merge(engteams, left_on="currentTeamId", right_on="wyId")
engplayer=engplayer.drop(columns=['birthArea','height','role','currentNationalTeamId','officialName','type','middleName','lastName','passportArea','weight','currentTeamId','firstName','foot'],axis=1)

engplayer=engplayer.drop(columns=['alpha3code','alpha2code','id','nation','wyId_y','teamname'])
engplayer=engplayer.rename(columns={"wyId_x": "PlayerId"})

final_player_passes=pd.merge(engplayer, final_passes, on='PlayerId')
final_player_passes

birthDate  PlayerId             shortName  AccuratePassesNumber  \
0    1987-04-24        48         J. Vertonghen                  1129   
1    1992-02-14        54            C. Eriksen                   869   
2    1990-09-19      8945           K. Trippier                   601   
3    1987-07-16     11152  M. Demb\u00e9l\u00e9                   558   
4    1994-01-15    210044               E. Dier                  1073   
..          ...       ...                   ...                   ...   
100  1991-04-19      9277               S. Cook                   438   
101  1995-02-18    134102           N. Ak\u00e9                   689   
102  1997-02-03    302518               L. Cook                   354   
103  1989-08-12      7929          T. Cleverley                   830   
104  1993-01-01     28292      A. Doucour\u00e9                  1010   

     TotalPassesNumber  
0                 2370  
1                 2196  
2                 1311  
3                 1365  
4                 2147  
..                 ...  
100               1241  
101               1539  
102               1184  
103               1081  
104               2178  

[105 rows x 5 columns]

In [59]:
Efficiency=[]
for i in range(len(final_player_passes.AccuratePassesNumber)):
    total=final_player_passes.TotalPassesNumber.iloc[i]
    accurate=final_player_passes.AccuratePassesNumber.iloc[i]
    Ratio=accurate/total
    Efficiency.append(Ratio)
Efficiency=pd.DataFrame(Efficiency)

In [60]:
final_player_passes.insert(4,'Efficiency',Efficiency)
final_player_ratio=final_player_passes
final_player_ratio

birthDate  PlayerId             shortName  AccuratePassesNumber  \
0    1987-04-24        48         J. Vertonghen                  1129   
1    1992-02-14        54            C. Eriksen                   869   
2    1990-09-19      8945           K. Trippier                   601   
3    1987-07-16     11152  M. Demb\u00e9l\u00e9                   558   
4    1994-01-15    210044               E. Dier                  1073   
..          ...       ...                   ...                   ...   
100  1991-04-19      9277               S. Cook                   438   
101  1995-02-18    134102           N. Ak\u00e9                   689   
102  1997-02-03    302518               L. Cook                   354   
103  1989-08-12      7929          T. Cleverley                   830   
104  1993-01-01     28292      A. Doucour\u00e9                  1010   

     Efficiency  TotalPassesNumber  
0      0.476371               2370  
1      0.395719               2196  
2      0.458429               1311  
3      0.408791               1365  
4      0.499767               2147  
..          ...                ...  
100    0.352941               1241  
101    0.447693               1539  
102    0.298986               1184  
103    0.767808               1081  
104    0.463728               2178  

[105 rows x 6 columns]

In [61]:
final_player_ratio.to_json(r'C:\Users\Debodeep\Documents\Sapienza Learning Materials\ADM\ADMHw2\final_player_ratio.json')

In [62]:
bestefficiency=final_player_ratio.nlargest(10, ['Efficiency']) 
bestefficiency=bestefficiency.reset_index()
bestefficiency=bestefficiency.drop(columns=['index','PlayerId'])
bestefficiency=bestefficiency.set_value(9, 'shortName', 'V.Lindelof')

bestefficiency

C:\Users\Debodeep\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  after removing the cwd from sys.path.


birthDate      shortName  AccuratePassesNumber  Efficiency  \
0  1989-08-12   T. Cleverley                   830    0.767808   
1  1986-08-20      A. Surman                   814    0.734657   
2  1986-02-26  Nacho Monreal                  1200    0.708383   
3  1994-05-28      J. Stones                   801    0.665282   
4  1985-09-10   L. Koscielny                  1056    0.633473   
5  1989-11-21       F. Delph                   970    0.612374   
6  1988-10-15   M. \u00d6zil                  1012    0.565363   
7  1992-02-21       P. Jones                   588    0.556291   
8  1986-01-08    David Silva                  1324    0.555835   
9  1990-12-26     V.Lindelof                   675    0.555556   

   TotalPassesNumber  
0               1081  
1               1108  
2               1694  
3               1204  
4               1667  
5               1584  
6               1790  
7               1057  
8               2382  
9               1215

### The above table shows the top 10 most efficient players. We can observe that the highest efficiency is 76.78% in the premier league dataset.